### PR TITLE
`fatxpool`: size limits implemented

### PR DESCRIPTION
**Pull request by michalkucharczyk**
This PR adds size-limits to the fork-aware transaction pool.

**Review Notes**
- Existing [`TrackedMap`](https://github.com/paritytech/polkadot-sdk/blob/58fd5ae4ce883f42c360e3ad4a5df7d2258b42fe/substrate/client/transaction-pool/src/graph/tracked_map.rs#L33-L41) is used in internal mempool to track the size of extrinsics:
https://github.com/paritytech/polkadot-sdk/blob/58fd5ae4ce883f42c360e3ad4a5df7d2258b42fe/substrate/client/transaction-pool/src/graph/tracked_map.rs#L33-L41

- In this PR, I also removed the logic that kept transactions in the `tx_mem_pool` if they were immediately dropped by the views. Initially, I implemented this as an improvement: if there was available space in the _mempool_ and all views dropped the transaction upon submission, the transaction would still be retained in the _mempool_.

  However, upon further consideration, I decided to remove this functionality to reduce unnecessary complexity. Now, when all views drop a transaction during submission, it is automatically rejected, with the `submit/submit_and_watch` call returning `ImmediatelyDropped`.

Closes: #5476

### CODE REVIEW

File: substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs [hunk 1]
```diff
@@ -164,15 +148,7 @@ where
 							.iter()
 							.all(|h| !self.stream_map.contains_key(h))
```
Comment (skunert) [inline substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs:149]: Can you remind me when this second condition here triggers? If I see it correctly it is about views keeping the tx valid but are not in the stream_map. Meaning they went inactive?

**Review by skunert — COMMENTED** [review]
I like the removal of the `initial_views`, I think the scenario was relatively niche :+1:

File: substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs (hunk 1 above)
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs:149]: Yes, exactly. Theoretically inactive views are no longer kept in the `dropped_listener`:
https://github.com/paritytech/polkadot-sdk/blob/9584dbda9ed5fc91b1837f35f401659239fdaff1/substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs#L422-L426

So this condition maybe some left-over (but it should not hurt us). I will take a closer look on it :)

**Review by skunert — APPROVED** [review]

File: substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs (hunk 1 above)
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs:149]: Actually this was check for silently dropped views. I will open a new PR with some improvements in this area.

File: substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs [hunk 2]
```diff
@@ -200,28 +225,44 @@ where
...
+	pub fn unwatched_and_watched_count(&self) -> (usize, usize) {
 		let transactions = self.transactions.read();
 		let watched_count = transactions.values().filter(|t| t.is_watched()).count();
 		(transactions.len() - watched_count, watched_count)
 	}
 
+	/// Returns the number of bytes used by all extrinsics in the the pool.
+	///
+	/// Intended for testing.
+	#[allow(dead_code)]
```
Comment (bkchr) [inline substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs:237]:
```suggestion
	#[cfg(test)]
```

?

File: substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs [hunk 3]
```diff
@@ -262,14 +297,8 @@ where
...
-		let mut transactions = self.transactions.write();
-		let hash = self.api.hash_and_length(&xt).0;
-		self.try_insert(
-			transactions.len(),
-			transactions.entry(hash),
-			hash,
-			TxInMemPool::new_watched(source, xt.clone()),
-		)
+		let (hash, lenght) = self.api.hash_and_length(&xt);
+		self.try_insert(hash, TxInMemPool::new_watched(source, xt.clone(), lenght))
```
Comment (bkchr) [inline substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs:301]:
```suggestion
		let (hash, length) = self.api.hash_and_length(&xt);
		self.try_insert(hash, TxInMemPool::new_watched(source, xt.clone(), length))
```

**Pull request merged**